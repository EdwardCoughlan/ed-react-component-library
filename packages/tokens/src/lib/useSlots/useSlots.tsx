/* eslint-disable import/prefer-default-export */
import {
  Children,
  createElement,
  Fragment,
  FunctionComponentElement,
  ReactFragment,
  ReactNode,
  ReactPortal,
} from 'react';

type Component2 = {
  props?: unknown;
  type?: () => { name?: string };
};
type CollectorType = {
  [keyof: string]: Array<ReactNode | ReactFragment | ReactPortal>;
};

type NamedSlot = {
  name?: string;
  children?: ReactNode;
};

export const useSlots = (
  slotChildren: ReactNode,
  namedSlots: string[],
) : [
    ({ name, children }: NamedSlot) => FunctionComponentElement<{ children?: ReactNode; }>,
    (slot: string) => boolean,
    (slot: string) => (ReactNode | ReactFragment | ReactPortal)[],
  ] => {
  const slots = Children.toArray(slotChildren).reduce(
    (collector: CollectorType, child) => {
      const component = child as Component2;

      const slotName = component?.type?.name && namedSlots.includes(component.type.name)
        ? component.type.name
        : 'children';

      if (!(slotName in collector)) {
        // eslint-disable-next-line no-param-reassign
        collector[slotName] = [];
      }
      collector[slotName].push(child);
      return collector;
    },
    { children: [] },
  );

  return [
    ({ name, children }: NamedSlot) : FunctionComponentElement<{ children?: ReactNode; }> => {
      if (!name) {
        return createElement(Fragment, null, slots.children);
      }
      const filteredChildren = name in slots ? slots[name] : children;
      return createElement(Fragment, null, filteredChildren);
    },
    (slot: string): boolean => slot in slots && !!slots[slot],
    (slot: string): (ReactNode | ReactFragment | ReactPortal)[] => slots[slot],
  ];
};
