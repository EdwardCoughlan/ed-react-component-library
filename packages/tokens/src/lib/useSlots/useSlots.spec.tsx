import React, { FC, PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import { useSlots } from './useSlots';

const AComponentForSlotting: FC = () => <div>A component for slotting</div>;

const ExampleSlottedComponent: FC<PropsWithChildren> = ({ children }) => {
  const [Slots, hasSlots] = useSlots(children, [AComponentForSlotting.name]);

  return (
    <div>
      {hasSlots(AComponentForSlotting.name) && (
        <Slots name={AComponentForSlotting.name} />
      )}
      <Slots name="NotFoundButHasDefault">
        <div>I am a default value</div>
      </Slots>
      <div>
        <Slots />
      </div>
    </div>
  );
};

test('Basic', () => {
  const { container } = render(
    <ExampleSlottedComponent>
      <AComponentForSlotting />
      <div>default item</div>
      and some string
    </ExampleSlottedComponent>,
  );
  expect(container).toMatchSnapshot();
});

test('Inverted', () => {
  const { container } = render(
    <ExampleSlottedComponent>
      <div>default item</div>
      and some string
      <AComponentForSlotting />
    </ExampleSlottedComponent>,
  );
  expect(container).toMatchSnapshot();
});

test('equivelant', () => {
  const render1 = render(
    <ExampleSlottedComponent>
      <div>default item</div>
      and some string
      <AComponentForSlotting />
    </ExampleSlottedComponent>,
  );
  const render2 = render(
    <ExampleSlottedComponent>
      <AComponentForSlotting />
      <div>default item</div>
      and some string
    </ExampleSlottedComponent>,
  );
  expect(render1.container.innerHTML).toBe(render2.container.innerHTML);
});
