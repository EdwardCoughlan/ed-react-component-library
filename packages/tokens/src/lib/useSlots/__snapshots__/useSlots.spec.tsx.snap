// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic 1`] = `
<div>
  <div>
    <div>
      A component for slotting
    </div>
    <div>
      I am a default value
    </div>
    <div>
      <div>
        default item
      </div>
      and some string
    </div>
  </div>
</div>
`;

exports[`Inverted 1`] = `
<div>
  <div>
    <div>
      A component for slotting
    </div>
    <div>
      I am a default value
    </div>
    <div>
      <div>
        default item
      </div>
      and some string
    </div>
  </div>
</div>
`;
